### ISSUE TITLE
Python   

### ISSUE THREAD

**Issue by Baibhavi-rgh** [Baibhavi-rgh/CSE-Project#1]
Python

import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# --- 1. Load Data (Simplified Structure for Example) ---
# Create a dummy DataFrame representing 'movies.csv'
data = {
    'movieId': [1, 2, 3, 4, 5],
    'title': ['Toy Story (1995)', 'Jumanji (1995)', 'Waiting to Exhale (1995)', 'Four Rooms (1995)', 'Father of the Bride Part II (1995)'],
    'genres': ['Adventure|Animation|Children|Comedy|Fantasy', 'Adventure|Children|Fantasy', 'Comedy|Drama', 'Comedy', 'Comedy'],
    'plot_keywords': ['toys|friendship', 'board game|magic', 'relationship|divorce', 'hotel|new year', 'family|pregnancy']
}
df = pd.DataFrame(data)

# --- 2. Feature Engineering: Create a "Soup" of Features ---
def create_soup(x):
    # Combine genres and plot keywords into a single string
    return ' '.join(x['genres'].split('|')) + ' ' + ' '.join(x['plot_keywords'].split('|'))

# Fill NaN/empty values just in case
df['genres'] = df['genres'].fillna('')
df['plot_keywords'] = df['plot_keywords'].fillna('')

df['soup'] = df.apply(create_soup, axis=1)

# --- 3. Vectorization (TF-IDF) ---
# TF-IDF stands for Term Frequency-Inverse Document Frequency.
# It measures how important a word is to a document (movie) in a collection (all movies).
tfidf = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf.fit_transform(df['soup'])

# --- 4. Similarity Calculation (Cosine Similarity) ---
# Calculate the cosine similarity matrix. This measures the angle between vectors.
# Smaller angle (closer to 1) means higher similarity.
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

# --- 5. Recommendation Function ---
# Create a mapping of movie title to index for easy lookup
indices = pd.Series(df.index, index=df['title']).drop_duplicates()

def get_content_recommendations(title, cosine_sim=cosine_sim, df=df, indices=indices):
    """Generates content-based recommendations."""
    if title not in indices:
        return f"Error: Movie '{title}' not found in the dataset."

    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwise similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores (descending)
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies (excluding itself)
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 recommended movie titles
    return df['title'].iloc[movie_indices]

# --- 6. Test the System ---
print("## Content-Based Recommendations for 'Toy Story (1995)'")
print(get_content_recommendations('Toy Story (1995)'))
